client/pov-recovery/src/lib.rs

```diff
@@ -343,13 +344,55 @@ where
 		self.import_block(block).await;
 	}
 
+	async fn chain_recovery(&mut self, mut hash: Block::Hash) {
+		let mut to_recover = Vec::new();
+		let do_recover = loop {
+			match self.parachain_client.block_status(&BlockId::Hash(hash)) {
+				Ok(BlockStatus::Unknown) => to_recover.push(hash),
+				Ok(_) => break true,
+				Err(e) => {
+					tracing::error!(
+						target: LOG_TARGET,
+						error = ?e,
+						block_hash = ?hash,
+						"Failed to get block status",
+					);
+					break false
+				},
+			}
+
+			hash = match self.pending_candidates.get(&hash) {
```

Comment:
Maybe I am missing something, but is it really an error if the hash is not in `pending_candidates`?

I see that you removed the removal of imported blocks, but it could still have been previously recovered and removed here no? https://github.com/paritytech/cumulus/blob/55520786c37f57600102ecfa6a008fb92add9a0e/client/pov-recovery/src/lib.rs#L233

Reply:
When we trigger an explicit recovery of block X we also want to silently trigger the recovery of all its ancestry.

To do this we need the block X ancestry hashes. This hashes are recovered from the `PendingCandidates` entries we have already saw.

If a parent of the block that we want to explicitly recover has never been announced by the RelayChain (I.e. is not in the pending candidates yet) then we are not able to perform the recovery at the moment